### DISCUSSION TITLE
Lesson 13: Upgradeable Smart Contracts. Question: We're get proxy contract's instance using the implementation contract instead of ERC1967Proxy contract.

### DISCUSSION THREAD

**Discussion by alfheimrShiven** [Cyfrin/foundry-full-course-cu#897] [General]
So in the `UpgradeBox.s.sol` we need to get an instance of the proxy contract in order to change it's implementation contract from BoxV1 to BoxV2. 
Refer line: https://github.com/Cyfrin/foundry-upgrades-f23/blob/main/script/UpgradeBox.s.sol#L27

Generally in order to get an instance of a contract whose address we have, we get it by `ContractName(address)` eg. `ERC20Mock(address(wEth))` where `wEth` contract inherits `ERC20Mock`, but in case of of getting the `ERC1967Proxy`  instance, we used the implementation contract `BoxV1` contract:

`BoxV1(payable(proxyAddress))` instead of using `ERC1967Proxy(proxyAddress)`. 

According to my understanding, since we need to call the `proxy.upgradeTo(address(newBox))` which is provided by `UUPSUpgradeable.sol` we are getting the proxy instance using the `BoxV1` contract which inherits `UUPSUpgradeable.sol`. But then shouldn't `UUPSUpgradeable.sol` be inheriting `ERC1967Proxy.sol` to return the `ERC1967Proxy` proxy instance? 

Basically my current understanding is that, `ERC1967Proxy` should somewhere be in the chain of inheritance of `BoxV1` contract in order for `BoxV1(payable(proxyAddress))` to return an instance of the `ERC1967Proxy`. Isn't it?

I'm sure my understanding is not correct somewhere, hence would need your view on this. Thanks.

**Comment by sunnyStefi** [comment]
Hi @alfheimrShiven! 

BoxV1 does not inherit from ERC1967Proxy, but from **UUPSUpgradeable**.

ERC1967Proxy it’s just used to store the **initial address of the proxy**, that’s immutable (the contract that users will always call). If you take a look at the code, inside ERC1967Proxy you can only get the address of the current implementation, not setting or working with other implementation contracts (aka upgrading).

In fact in the code you mentioned you can also call directly the UUPSUpgradeable contract:

```solidity
//updated for open zeppelin version 5. For version 4 you can keep using upgradeTo(address)
UUPSUpgradeable contractToBeUpgraded = UUPSUpgradeable(payable(proxy));
proxy.upgradeToAndCall(contractToBeUpgradedWith, new bytes(0));
``` 

Both contracts (ERC1967Proxy and UUPSUpgradeable) refer to ERC1967Utils library to get and set contract proxy implementations.

Hope it helps, otherwise pls let me know!
Stefi

**Reply by cromewar** [reply]
Does this works for you @alfheimrShiven?

**Reply by alfheimrShiven** [reply]
Yes I got it @cromewar 
Thanks @sunnyStefi for pointing out that the proxy contract instance is being returned from `UUPSUpgradable`. Now I understand.
